eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -238,7 +239,7 @@ deriving via
   instance
     NoThunks PV1.EvaluationContext
```

Comment:
@nfrisby Is this hand-written instance problematic?

Reply:
`EvaluationContext` is declared in the `plutus` repo. It does have a `Generic` instance, so you could again drop the `InspectHeap` usage here.

The overall intuition is that `InspectHeap` is an escape hatch for use when the `NoThunks` instances get rough/inconvenient. See https://hackage.haskell.org/package/nothunks-0.1.3/docs/NoThunks-Class.html#t:InspectHeap

I'm not familiar with all the types that occur within the `EvaluationContext` -- maybe one of them deep inside it does require `InspectHeap`? (Perhaps `BuiltinsRuntime`?)

Ideally, the `plutus` repo would be declaring the `NoThunks` instances for their data types. Using `InspectHeap` as a stop-gap measure until the instances are upstreamed doesn't seem crazy to me.

Reply:
Many thanks for the explanation. When I first added the `EvaluationContext` to the cost models, I was not able to automatically derive these instances, but I've since forgotten the details (" should have left a trail of comments!). I'll go back with renewed confidence!

Reply:
ok, here is the error that I encounter when trying to make a `NoThunks` instance automatically with `instance NoThunks PV1.EvaluationContext`:

```
src/Cardano/Ledger/Alonzo/Scripts.hs:237:10: error:
    * No instance for (NoThunks
                         (PlutusCore.Evaluation.Machine.MachineParameters.MachineParameters
                            plutus-core-1.0.0.0:UntypedPlutusCore.Evaluation.Machine.Cek.CekMachineCosts.CekMachineCosts
                            UntypedPlutusCore.Evaluation.Machine.Cek.Internal.CekValue
                            plutus-core-1.0.0.0:PlutusCore.Default.Universe.DefaultUni
                            plutus-core-1.0.0.0:PlutusCore.Default.Builtins.DefaultFun))
        arising from a use of `NoThunks.Class.$dmwNoThunks'
    * In the expression:
        NoThunks.Class.$dmwNoThunks @(EvaluationContext)
      In an equation for `wNoThunks':
          wNoThunks = NoThunks.Class.$dmwNoThunks @(EvaluationContext)
      In the instance declaration for `NoThunks EvaluationContext'
    |
237 | instance NoThunks PV1.EvaluationContext
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So it looks like `MachineParameters` is the culprit. I'll make an issue in the Plutus repo, and leave a TODO in the ledger code.

Reply:
I'll go with the workaround for `EvaluationContext` for now, and we can remove it after https://github.com/input-output-hk/plutus/issues/4687 is resolved.